### PR TITLE
Initial version of BigQuery query execution component.

### PR DESCRIPTION
**Pull request by cbreuel**
This is a component that runs a query in BigQuery and stores the results in a new table, and optionally also exports the results to GCS.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/601)

### CODE REVIEW

File: components/bigquery/query/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+FROM ml-pipeline-bigquery-base
+
+RUN apt-get update -y && \
```
Comment (qimingj) [inline components/bigquery/query/Dockerfile]: Can you add some comments why we need to install and then purge build-essential? If it is needed can these be moved to base, so keep only the source code and entrypoint in this Dockerfile?

File: components/bigquery/query/src/query.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+      type=str,
+      required=False,
+      help='GCS URL where results will be saved as a CSV.')
+  parser.add_argument(
+      '--query',
+      type=str,
+      required=True,
+      help='The SQL query to be run in BigQuery')
+  parser.add_argument(
+      '--dataset_id', type=str, help='Dataset of the destination table.')
```
Comment (qimingj) [inline components/bigquery/query/src/query.py]: Both dataset_id and table_id are required?

File: components/bigquery/query/src/query.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+      required=True,
+      help='The GCP project to run the query.')
+
+  args = parser.parse_args()
+  return args
+
+
+def run_query(query, project_id, dataset_id, table_id, output):
+  results = _query(query, project_id, dataset_id, table_id, output)
+  results['output'] = output
```
Comment (qimingj) [inline components/bigquery/query/src/query.py]: It seems useful if we store the files pattern, not just the output dir into results. As the component provider, I probably know the file pattern beforehand, and that is useful information to component consumers.

File: components/bigquery/query/src/query/setup.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
+# Copyright 2018 Google LLC
```
Comment (qimingj) [inline components/bigquery/query/src/query/setup.py]: Is setup.py needed?

File: components/bigquery/query/Dockerfile (hunk 1 above)
Comment (cbreuel) [reply components/bigquery/query/Dockerfile]: I copied this from [the tfdv component](https://github.com/kubeflow/pipelines/blob/master/components/dataflow/tfdv/Dockerfile), so to be honest I'm not sure if it's needed. I will try removing it.

File: components/bigquery/query/src/query.py (hunk 2 above)
Comment (cbreuel) [reply components/bigquery/query/src/query.py]: Yes, the table will be created/updated inside the dataset.

File: components/bigquery/query/src/query/setup.py (hunk 4 above)
Comment (cbreuel) [reply components/bigquery/query/src/query/setup.py]: Looks like it wasn't necessary, removed.

File: components/bigquery/query/Dockerfile (hunk 1 above)
Comment (cbreuel) [reply components/bigquery/query/Dockerfile]: Removed.

**Comment by cbreuel** [comment]
Made some changes based on your comments, PTAL.

File: components/bigquery/query/src/query.py (hunk 2 above)
Comment (qimingj) [reply components/bigquery/query/src/query.py]: then "required=True"?

File: components/bigquery/query/src/query.py (hunk 3 above)
Comment (qimingj) [reply components/bigquery/query/src/query.py]: Here the output is just whatever is passed in right? Is it a dir or a file prefix? I would prefer the output includes the pattern or file (if single file). For example, if passed in output is "gs://b/dir", the output could be "gs://b/dir/files-*"?

Comment (qimingj) [reply components/bigquery/query/src/query.py]: /ok-to-test
/lgtm

**Comment by qimingj** [comment]
You are right. Based on https://google-cloud.readthedocs.io/en/latest/bigquery/generated/google.cloud.bigquery.client.Client.extract_table.html output should be the file path.

**Pull request merged**